File: 2018-kakao-blind-recruitment/Q12.자동완성.py
class Trie:
    def __init__(self):
        self.head = {}

    def add(self, word):
        curr = self.head
        for ch in word:
            if ch not in curr:
                curr[ch] = {}
            curr = curr[ch]

            if 'child' not in curr:
                curr['child'] = 1
            else:
                curr['child'] += 1

    def query(self, word):
        curr = self.head
        cnt = 0

        for ch in word:
            curr = curr[ch]
            cnt += 1

            if curr['child'] == 1:
                break

        return cnt


def solution(words):
    answer = 0
    trie = Trie()
    for word in words:
        trie.add(word)

    for word in words:
        answer += trie.query(word)

    return answer

File: 2018-kakao-blind-recruitment/Q12.자동완성.js
class Trie {
	#head = new Map();

	constructor() {}

	insert(word) {
		let curr = this.#head;

		for (const ch of word) {
			if (!curr.has(ch)) curr.set(ch, new Map());
			curr = curr.get(ch);

			if (!curr.has('child')) curr.set('child', 0);
			curr.set('child', curr.get('child') + 1);
		}
	}

	query(word) {
		let cnt = 0;
		let curr = this.#head;

		for (const ch of word) {
			curr = curr.get(ch);
			cnt += 1;

			if (curr.get('child') === 1) break;
		}

		return cnt;
	}
}

function solution(words) {
	const trie = new Trie();

	words.forEach(word => trie.insert(word));

	return words.map(word => trie.query(word)).reduce((a, b) => a + b, 0);
}

solution(['go', 'gone', 'guild']);
